benchmarks/src/copy.cpp

```diff
@@ -0,0 +1,117 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <algorithm>
+#include <benchmark/benchmark.h>
+#include <stddef.h>
+#include <string.h>
+#include <vector>
+
+#include <udt.hpp>
+
+template <typename Contained>
+static void handwritten_loop(benchmark::State& state) {
+    const size_t r0 = static_cast<size_t>(state.range(0));
+    std::vector<Contained> in_buffer(r0);
+    std::vector<Contained> out_buffer(r0);
+    for ([[maybe_unused]] auto _ : state) {
+        benchmark::DoNotOptimize(in_buffer);
+        Contained* in_ptr     = in_buffer.data();
+        Contained* in_ptr_end = in_ptr + r0;
+        Contained* out_ptr    = out_buffer.data();
+        while (in_ptr != in_ptr_end) {
+            *out_ptr++ = *in_ptr++;
+        }
+
+        benchmark::DoNotOptimize(out_buffer);
+    }
+}
+
+template <typename Contained>
+static void handwritten_loop_n(benchmark::State& state) {
+    const size_t r0 = static_cast<size_t>(state.range(0));
+    std::vector<Contained> in_buffer(r0);
+    std::vector<Contained> out_buffer(r0);
+    for ([[maybe_unused]] auto _ : state) {
+        benchmark::DoNotOptimize(in_buffer);
+        Contained* in_ptr  = in_buffer.data();
+        Contained* out_ptr = out_buffer.data();

```

Comment:
Similarly, `in_ptr` could be `const Contained* const` and `out_ptr` could be `Contained* const`.